src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -97,9 +98,24 @@ private void getMealList(HttpServletRequest request, HttpServletResponse respons
         // Every object Meal adds if it contains at least one keyword in at least one field 
         // (type, title, description, ingredients).
         List<Meal> searchedMeal = new ArrayList<>();
+        HashMap<Integer, List<Meal>> sortedMeal = new HashMap<>();
         for (Meal meal : meals) {
-            if (isResultOfSearch(meal, params)) {
-                searchedMeal.add(meal);
+            int frequency = getFrequency(meal, params);
+            if (frequency > 0) {
+                List<Meal> list = new ArrayList<>();
+                if (sortedMeal.get(frequency) != null) {
+                    list = sortedMeal.get(frequency);
+                }
+                list.add(meal);
+                sortedMeal.put(frequency, list);
+            }
+        }
+        for (int i = params.size(); i > 0; i --) {
```

Comment:
If you had a `TreeMap` instead of a `HashMap` here, you would be able to just go through the map using `.descendingMap()` or `.descendingKeySet()`.
`TreeMap` keeps all elements in order (i.e. if you put entries under keys `1`, `7`, `5` and `9`, then they would be sorted in ascending order in the map, unlike the pseudo-random order of a `HashMap`) at a cost of making all insert and get operations in `O(log(n))` time instead of `O(1)`. These structures are great for sorting operations like this search result sorting.
If you have the time, please consider trying `TreeMap` here instead of `HashMap` (and then replacing the `for (int i = ....)` with a loop over `mymap.descendingMap()`, because otherwise the whole endeavour won't have any benefits; also looping over the map guarantees the value can't be `null`, so one less check to make!).